pkg/apis/serving/fieldmask.go

```diff
@@ -50,12 +50,89 @@ func VolumeSourceMask(in *corev1.VolumeSource) *corev1.VolumeSource {
 	// Allowed fields
 	out.Secret = in.Secret
 	out.ConfigMap = in.ConfigMap
+	out.Projected = in.Projected
 
 	// Too many disallowed fields to list
 
 	return out
 }
 
+// VolumeProjectionMask performs a _shallow_ copy of the Kubernetes VolumeProjection
+// object to a new Kubernetes VolumeProjection object bringing over only the fields allowed
+// in the Knative API. This does not validate the contents or the bounds of the provided fields.
+func VolumeProjectionMask(in *corev1.VolumeProjection) *corev1.VolumeProjection {
+	if in == nil {
+		return nil
+	}
+
+	out := new(corev1.VolumeProjection)
+
+	// Allowed fields
+	out.Secret = in.Secret
+	out.ConfigMap = in.ConfigMap
+
+	// Disallowed fields
+	// This list is unnecessary, but added here for clarity
+	out.DownwardAPI = nil
+	out.ServiceAccountToken = nil
+
+	return out
+}
+
+// ConfigMapProjectionMask performs a _shallow_ copy of the Kubernetes ConfigMapProjection
+// object to a new Kubernetes ConfigMapProjection object bringing over only the fields allowed
+// in the Knative API. This does not validate the contents or the bounds of the provided fields.
+func ConfigMapProjectionMask(in *corev1.ConfigMapProjection) *corev1.ConfigMapProjection {
+	if in == nil {
+		return nil
+	}
+
+	out := new(corev1.ConfigMapProjection)
+
+	// Allowed fields
+	out.LocalObjectReference = in.LocalObjectReference
+	out.Items = in.Items
+	out.Optional = in.Optional
+
+	return out
+}
+
+// SecretProjectionMask performs a _shallow_ copy of the Kubernetes SecretProjection
+// object to a new Kubernetes SecretProjection object bringing over only the fields allowed
+// in the Knative API. This does not validate the contents or the bounds of the provided fields.
+func SecretProjectionMask(in *corev1.SecretProjection) *corev1.SecretProjection {
+	if in == nil {
+		return nil
+	}
+
+	out := new(corev1.SecretProjection)
+
+	// Allowed fields
+	out.LocalObjectReference = in.LocalObjectReference
+	out.Items = in.Items
```

Comment:
Items and Optional are not exercised in the e2e tests and will drop our API coverage. It would be good to have tests covering the usage of these fields if we want to support them.

Reply:
Added coverage of `Items` for both tests to the e2e.

We don't cover Optional anywhere and should probably consider whether supporting it at all makes sense (for now).